br/pkg/lightning/backend/external/engine.go

```diff
@@ -146,12 +161,236 @@ func (e *Engine) getAdjustedConcurrency() int {
+	}
+	endOffs, err := seekPropsOffsets(ctx, endKey, statsFiles, storage, false)
+	if err != nil {
+		return nil, nil, err
+	}
+	for i := range statsFiles {
+		result[i] = (endOffs[i] - startOffs[i]) / uint64(ConcurrentReaderBufferSizePerConc)
+		if result[i] < 16 {
+			result[i] = 1
+		} else {
+			logutil.Logger(ctx).Info("found hotspot file in checkConcurrentFiles",
+				zap.String("filename", statsFiles[i]),
+			)
+		}
+	}
+	return result, startOffs, nil
+}
+
+func readOneFile(
+	ctx context.Context,
+	storage storage.ExternalStorage,
+	dataFile string,
+	startKey, endKey []byte,
+	startOffset uint64,
+	concurrency uint64,
+	bufPool *membuf.Pool,
+	output *memKVsAndBuffers,
+) error {
+	readAndSortDurHist := metrics.GlobalSortReadFromCloudStorageDuration.WithLabelValues("read_one_file")
+
+	ts := time.Now()
+
+	rd, err := newKVReader(ctx, dataFile, storage, startOffset, 64*1024)
+	if err != nil {
+		return err
+	}
+	defer rd.Close()
+	if concurrency > 1 {
+		rd.byteReader.enableConcurrentRead(
+			storage,
+			dataFile,
+			int(concurrency)*2,
+			ConcurrentReaderBufferSizePerConc,
+			bufPool.NewBuffer(),
+		)
+		err = rd.byteReader.switchConcurrentMode(true)
+		if err != nil {
+			return err
+		}
+	}
+
+	// this buffer is associated with data slices and will return to caller
+	memBuf := bufPool.NewBuffer()
+	keys := make([][]byte, 0, 1024)
+	values := make([][]byte, 0, 1024)
+
+	var prevKey []byte
+	for {
+		k, v, err := rd.nextKV()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if bytes.Compare(k, startKey) < 0 {
+			continue
+		}
+		if bytes.Compare(k, endKey) >= 0 {
+			break
+		}
+		if prevKey != nil && bytes.Compare(prevKey, k) >= 0 {
+			log.FromContext(ctx).Error("kv is not in increasing order", zap.ByteString("prevKey", prevKey), zap.ByteString("key", k))
+		}
+		prevKey = slices.Clone(k)
+		// TODO(lance6716): we are copying every KV from rd's buffer to memBuf, can we
+		// directly read into memBuf?
+		keys = append(keys, memBuf.AddBytes(k))
+		values = append(values, memBuf.AddBytes(v))
+	}
+	readAndSortDurHist.Observe(time.Since(ts).Seconds())
+	output.mu.Lock()
+	output.keys = append(output.keys, keys...)
+	output.values = append(output.values, values...)
+	output.memKVBuffers = append(output.memKVBuffers, memBuf)
+	output.mu.Unlock()
+	return nil
+}
+
+func readAllData(
+	ctx context.Context,
+	storage storage.ExternalStorage,
+	dataFiles, statsFiles []string,
+	startKey, endKey []byte,
+	bufPool *membuf.Pool,
+	output *memKVsAndBuffers,
+) error {
+	logutil.Logger(ctx).Info("enter readAllData",
+		zap.Int("data-file-count", len(dataFiles)),
+		zap.Int("stat-file-count", len(statsFiles)),
+		zap.Binary("start-key", startKey),
+		zap.Binary("end-key", endKey),
+	)
+
+	concurrences, startOffsets, err := checkConcurrentFiles(
+		ctx,
+		storage,
+		statsFiles,
+		startKey,
+		endKey,
+	)
+	if err != nil {
+		return err
+	}
+	var eg errgroup.Group
+	for i := range dataFiles {
```

Comment:
Why read all data files concurrently, In the worst case the concurrency will be so large.